### ISSUE TITLE
Murmur filter does not detect failures reading systemd-journal

### ISSUE THREAD

**Issue by S3bast1** [fail2ban/fail2ban#2043]
Hi,

I got a problem with the jail "murmur" of fail2ban.
The systemd journal detects fail attempts but fail2ban didn't recognize it (no message added in fail2ban.log).

Below the description of the system used, config and the log event in systemd.
Anyone got an idea please ? Thanks.

Fail2ban 0.10.2
Backend : systemd

Journalctl show the log error on murmurd and display "Invalid password server" :
```
_murmurd[1461]: <W>2018-02-08 13:02:19.512 1 => <21:Client1(-1)> Rejected connection from zzz.zzz.zzz.zzz:z: Invalid server password_
```
Jail config :
```INI
[murmur]

enabled = true
port     = 64738
action   = %(banaction)s[name=%(__name__)s-tcp, port="%(port)s", protocol=tcp, chain="%(chain)s", actname=%(banaction)s-tcp]
           %(banaction)s[name=%(__name__)s-udp, port="%(port)s", protocol=udp, chain="%(chain)s", actname=%(banaction)s-udp]
```
Filter murmur (default settings) :
```INI
[INCLUDES]
before = common.conf
[Definition]
_daemon = murmurd
_usernameregex = [^>]+
_prefix = \s+\d+ => <\d+:%(_usernameregex)s\(-1\)> Rejected connection from <HOST>:\d+:
prefregex = ^%(_prefix)s <F-CONTENT>.+</F-CONTENT>$
failregex = ^Invalid server password$
            ^Wrong certificate or password for existing user$
ignoreregex =
datepattern = ^<W>{DATE}
```

**Comment by sebres** [comment]
What exactly journalctl shows you? I mean is the date-time string `<W>2018-02-08 13:02:19.512` rather a part of message, written from murmurd or it was date-time of the event which `journalctl` just output...
So do you see still one date-time additionally (double time-pattern)?
Can you provide any log-line from other service, e. g. from sshd. For example my output from sshd via journalctl:
```
Feb 08 18:53:25 srv sshd[12183]: pam_unix(sshd:session): session opened for user guest by (uid=123)
```

**Comment by S3bast1** [comment]
The full lines from journalctl :

févr. 08 13:02:19 vps289912.ovh.net murmurd[1461]: <W>2018-02-08 13:02:19.022 1 => <21:(-1)> New connection: zzz.zzz.zzz.zzz:yyyyy
févr. 08 13:02:19 vps289912.ovh.net murmurd[1461]: <W>2018-02-08 13:02:19.509 1 => <21:(-1)> Client version 1.2.5 (Android: Plumble Free 3.2.0)
févr. 08 13:02:19 vps289912.ovh.net murmurd[1461]: <W>2018-02-08 13:02:19.512 1 => <21:Client1(-1)> Rejected connection from zzz.zzz.zzz.zzz:yyyyy Invalid server password
févr. 08 13:02:19 vps289912.ovh.net murmurd[1461]: <W>2018-02-08 13:02:19.516 1 => <21:Client1	(-1)> Connection closed:  [-1]

**Comment by sebres** [comment]
So it looks like dual date-time issue...

Just to be sure, to help find it, can you provide some failure lines (3-5 lines) from:
```bash
fail2ban-regex -l heavydebug "systemd-journal" murmur | grep "Read systemd journal entry:"
```

**Comment by S3bast1** [comment]
The problem should be here :


```
fail2ban-regex -l heavydebug "systemd-journal" murmur | grep "Read systemd journal entry:"
Traceback (most recent call last):
  File "/usr/bin/fail2ban-regex", line 34, in <module>
    exec_command_line()
  File "/usr/lib/python3.6/site-packages/fail2ban/client/fail2banregex.py", line 685, in exec_command_line
    if not fail2banRegex.start(args):
  File "/usr/lib/python3.6/site-packages/fail2ban/client/fail2banregex.py", line 617, in start
    output( "Use    journal match : %s" % " ".join(journalmatch) )
```
**TypeError: can only join an iterable**

**Comment by S3bast1** [comment]
And also tried this before :

fail2ban-regex "systemd-journal" /etc/fail2ban/filter.d/murmur.conf

Running tests
=============

Use   failregex filter file : murmur, basedir: /etc/fail2ban
Use      datepattern : Default Detectors
Use         systemd journal
Use         encoding : UTF-8
Traceback (most recent call last):
  File "/usr/bin/fail2ban-regex", line 34, in <module>
    exec_command_line()
  File "/usr/lib/python3.6/site-packages/fail2ban/client/fail2banregex.py", line 685, in exec_command_line
    if not fail2banRegex.start(args):
  File "/usr/lib/python3.6/site-packages/fail2ban/client/fail2banregex.py", line 617, in start
    output( "Use    journal match : %s" % " ".join(journalmatch) )
TypeError: can only join an iterable

It's the same error :(

**Comment by sebres** [comment]
Is another bug (of fail2ban-regex). I'll fix it later. 
This occurs, because filter does not have `journalmatch` specified (also this is not so good for usage in production).

To check it works, can you please temporary change your murmur-filter (with replace of _prefix and append journalmatch after datepattern):
```diff
-_prefix = \s+\d+ => <\d+:%(_usernameregex)s\(-1\)> Rejected connection from <HOST>:\d+:
+_prefix = (?:\<W\>[\d\-]+ [\d:]+.\d+)?\s+\d+ => <\d+:%(_usernameregex)s\(-1\)> Rejected connection from <HOST>:\d+:
...
datepattern = ^<W>{DATE}
+ journalmatch = _SYSTEMD_UNIT=murmurd.service + _COMM=murmurd
```
And re-try my suggested test above.

**Comment by S3bast1** [comment]
Thanks Sebres.

I still have an error... the filter is now :

```
[INCLUDES]

before = common.conf
[Definition]
_daemon = murmurd
_usernameregex = [^>]+

_prefix = (?:\<W\>[\d\-]+ [\d:]+.\d+)?\s+\d+ => <\d+:%(_usernameregex)s\(-1\)> Rejected connection from <HOST>:\d+:

prefregex = ^%(_prefix)s <F-CONTENT>.+</F-CONTENT>$

failregex = ^Invalid server password$
            ^Wrong certificate or password for existing user$

ignoreregex =

datepattern = ^<W>{DATE}
 journalmatch = _SYSTEMD_UNIT=murmurd.service + _COMM=murmurd

```

And get :

```
fail2ban-regex -l heavydebug "systemd-journal" murmur | grep "Read systemd journal entry:"
Traceback (most recent call last):
  File "/usr/bin/fail2ban-regex", line 34, in <module>
    exec_command_line()
  File "/usr/lib/python3.6/site-packages/fail2ban/client/fail2banregex.py", line 685, in exec_command_line
    if not fail2banRegex.start(args):
  File "/usr/lib/python3.6/site-packages/fail2ban/client/fail2banregex.py", line 617, in start
    output( "Use    journal match : %s" % " ".join(journalmatch) )
TypeError: can only join an iterable
```

**Comment by sebres** [comment]
you should remove space before journalmatch

**Comment by S3bast1** [comment]
Thanks Sebres, now when I launch you're test, I got the lines of murmurd failed login from systemd.

But... Fail2ban don't detect them (no lines added in fail2ban.log).
I tried again the command :

fail2ban-regex "systemd-journal" /etc/fail2ban/filter.d/murmur.conf

```
Running tests
=============

Use   failregex filter file : murmur, basedir: /etc/fail2ban
Use      datepattern : Default Detectors
Use         systemd journal
Use         encoding : UTF-8
Use    journal match : _SYSTEMD_UNIT=murmurd.service + _COMM=murmurd


Results
=======

Failregex: 0 total

Ignoreregex: 0 total

Date template hits:

Lines: 1048 lines, 0 ignored, 0 matched, 1048 missed
[processed in 0.19 sec]

Missed line(s): too many to print.  Use --print-all-missed to print all 1048 lines

```

No failed login are detected... have you an idea please ?

**Comment by sebres** [comment]
> No failed login are detected... have you an idea please ?

Possibly still an "error" in the failregex.

I told to try this (with debug and grep any failure line from there):
```
fail2ban-regex -l heavydebug "systemd-journal" murmur | grep "Read systemd journal entry:"
```
I need to see this failure-line, how it will be really retrieved from the systemd-journal.

**Comment by S3bast1** [comment]
That's an exemple of connection failed (got with you're test command) :

```
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.059998 zzzzzzzzzzzz murmurd[1461]: <W>2018-02-09 09:55:12.059 1 => <27:(-1)> New connection: zzz.zzz.zzz.zzz:zzzzz
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.391748 zzzzzzzzzzzz murmurd[1461]: <W>2018-02-09 09:55:12.391 1 => <27:(-1)> Client version 1.2.5 (Android: Plumble Free 3.2.0)
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.395962 zzzzzzzzzzzz murmurd[1461]: <W>2018-02-09 09:55:12.395 1 => <27:Client1(-1)> Rejected connection from zzz.zzz.zzz.zzz:zzzzz: Invalid server password
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.399920 zzzzzzzzzzzz murmurd[1461]: <W>2018-02-09 09:55:12.399 1 => <27:Client1(-1)> Connection closed:  [-1]
```

**Comment by sebres** [comment]
So can you try this (whether it will match without debug):
```INI
# Fail2Ban filter for murmur/mumble-server
#

[Definition]

_daemon = murmurd

# N.B. If you allow users to have usernames that include the '>' character you
#      should change this to match the regex assigned to the 'username'
#      variable in your server config file (murmur.ini / mumble-server.ini).
_usernameregex = [^>]+

# Prefix for systemd-journal (with second date-pattern as optional match):
#
__prefix_journal = (?:\s+\S+\s+%(_daemon)s\[\d+\]:(?:\s+\<W\>[\d\-]+ [\d:]+.\d+)?)

__prefix_line = %(__prefix_journal)s?

_prefix = %(__prefix_line)s\s+\d+ => <\d+:%(_usernameregex)s\(-1\)> Rejected connection from <HOST>:\d+:

prefregex = ^%(_prefix)s <F-CONTENT>.+</F-CONTENT>$

failregex = ^Invalid server password$
            ^Wrong certificate or password for existing user$

ignoreregex =

datepattern = ^<W>{DATE}

journalmatch = _SYSTEMD_UNIT=murmurd.service + _COMM=murmurd

# DEV Notes:
#
# Author: Ross Brown
```

**Comment by S3bast1** [comment]
:(( It's still ignoring the failed in murmur :


```
fail2ban-regex "systemd-journal" /etc/fail2ban/filter.d/murmur.conf

Running tests
=============

Use   failregex filter file : murmur, basedir: /etc/fail2ban
Use      datepattern : Default Detectors
Use         systemd journal
Use         encoding : UTF-8
Use    journal match : _SYSTEMD_UNIT=murmurd.service + _COMM=murmurd


Results
=======

Failregex: 0 total

Ignoreregex: 0 total

Date template hits:

Lines: 1048 lines, 0 ignored, 0 matched, 1048 missed
[processed in 0.38 sec]

Missed line(s): too many to print.  Use --print-all-missed to print all 1048 lines
```



Result with you're debug command :
```
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.059998 zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.059 1 => <27:(-1)> New connection: zzz.zzz.zzz.zzz:zzzzz
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.391748 zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.391 1 => <27:(-1)> Client version 1.2.5 (Android: Plumble Free 3.2.0)
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.395962 zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.395 1 => <27:Client1(-1)> Rejected connection from zzz.zzz.zzz.zzz:zzzzz: Invalid server password
H: [~jailless~] Read systemd journal entry: 2018-02-09T09:55:12.399920 zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.399 1 => <27:Client1(-1)> Connection closed:  [-1]
```

**Comment by sebres** [comment]
Can you please show the same, but with  ` | grep "Looking for match of "`

**Comment by S3bast1** [comment]
Sure, the result :

```
fail2ban-regex -l heavydebug "systemd-journal" murmur | grep "Looking for match of "
H: Looking for match of [('', '2018-02-09T09:55:12.059998', 'zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.059 1 => <27:(-1)> New connection: zzz.zzz.zzz.zzz:zzzzz')]
H: Looking for match of [('', '2018-02-09T09:55:12.391748', 'zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.391 1 => <27:(-1)> Client version 1.2.5 (Android: Plumble Free 3.2.0)')]
H: Looking for match of [('', '2018-02-09T09:55:12.395962', 'zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.395 1 => <27:Client1(-1)> Rejected connection from zzz.zzz.zzz.zzz:zzzzz: Invalid server password')]
H: Looking for match of [('', '2018-02-09T09:55:12.399920', 'zzz.zzz.zzz murmurd[1461]: <W>2018-02-09 09:55:12.399 1 => <27:Client1(-1)> Connection closed:  [-1]')]
```

**Comment by sebres** [comment]
```diff
-__prefix_journal = (?:\s+\S+\s+%(_daemon)s\[\d+\]:(?:\s+\<W\>[\d\-]+ [\d:]+.\d+)?)
+__prefix_journal = (?:\S+\s+%(_daemon)s\[\d+\]:(?:\s+\<W\>[\d\-]+ [\d:]+.\d+)?)
```
I've corrected also the filter-sample in comment above.

**Comment by S3bast1** [comment]
Seems to be better :)) :

fail2ban-regex "systemd-journal" /etc/fail2ban/filter.d/murmur.conf

```
Running tests
=============

Use   failregex filter file : murmur, basedir: /etc/fail2ban
Use      datepattern : Default Detectors
Use         systemd journal
Use         encoding : UTF-8
Use    journal match : _SYSTEMD_UNIT=murmurd.service + _COMM=murmurd


Results
=======

Failregex: 31 total
|-  #) [# of hits] regular expression
|   1) [31] ^Invalid server password$
`-

Ignoreregex: 0 total

Date template hits:

Lines: 1048 lines, 0 ignored, 31 matched, 1017 missed
[processed in 0.36 sec]

Missed line(s): too many to print.  Use --print-all-missed to print all 1017 lines
```

Will try another failed attempt :)

**Comment by S3bast1** [comment]
Happy to tells you that it works fine :))))

`2018-02-09 11:19:37,500 fail2ban.actions        [6570]: NOTICE  [murmur] Ban zzz.zzz.zzz.zzz`

Many thanks Sebres :))))

**Comment by sebres** [comment]
Glad it works, I'll fix it later in the mainstream branches.
Thank you.

**Comment by S3bast1** [comment]
Happy to help for Fail2ban debug :)

**Comment by sebres** [comment]
Fixed in 19a5a2f8c03d2c3b1fd3e7622db12619a1488ade.
